cpp/net/connection_pool.cc

```diff
@@ -30,20 +132,57 @@ const HostPortPair& ConnectionPool::Connection::other_end() const {
 
 
 ConnectionPool::ConnectionPool(libevent::Base* base)
-    : base_(CHECK_NOTNULL(base)), cleanup_scheduled_(false) {
+    : base_(CHECK_NOTNULL(base)),
+      cleanup_scheduled_(false),
+      ssl_ctx_(CHECK_NOTNULL(SSL_CTX_new(TLSv1_client_method())),
+               SSL_CTX_free) {
+  // Try to load trusted root certificates.
+  // TODO(alcutter): This is probably Linux specific, we'll need other sections
+  // for OSX etc.
+  if (SSL_CTX_load_verify_locations(ssl_ctx_.get(),
+                                    FLAGS_trusted_root_certs.c_str(),
+                                    nullptr) != 1) {
+    DumpOpenSSLErrorStack();
+    LOG(FATAL) << "Couldn't load trusted root certificates.";
+  }
+
+  SSL_CTX_set_verify(ssl_ctx_.get(), SSL_VERIFY_PEER,
+                     Connection::SSLVerifyCallback);
 }
 
 
 // static
-evhtp_res ConnectionPool::Connection::ConnectionClosedHook(
-    evhtp_connection_t* conn, void* arg) {
+evhtp_res ConnectionPool::Connection::ConnectionErrorHook(
+    evhtp_connection_t* conn, evhtp_error_flags errtype, void* arg) {
   CHECK_NOTNULL(conn);
   CHECK_NOTNULL(arg);
+  CHECK(libevent::Base::OnEventThread());
   ConnectionPool::Connection* const c(
       static_cast<ConnectionPool::Connection*>(arg));
   VLOG(1) << "Releasing connection to " << c->other_end().first << ":"
           << c->other_end().second;
+
   CHECK_EQ(conn, c->connection());
+
+  if (conn->ssl) {
+    {
+      lock_guard<mutex> lock(ssl_mutex_);
+      CHECK_EQ(1, connections_by_ssl_.erase(conn->ssl));
```

Comment:
This is the only spot this is called, and it's in a method called "error hook", so I'm wondering who cleans this up when there's no error? :wink: 

Cleaning this up in the `ConnectionPool::Connection` destructor would be logical, as this is added in its constructor.